tangelo/toolboxes/molecular_computation/rdms.py

```diff
@@ -41,6 +48,96 @@ def matricize_2rdm(two_rdm, n_orbitals):
+
+    # Check if the data dict is expectation values or raw frequency data in the {basis: hist} format
+    if isinstance(exp_data, dict) and isinstance(list(exp_data.values())[0], float):
+        exp_vals = {pauli_string_to_of(term): exp_val for term, exp_val in exp_data.items()}
+    elif isinstance(exp_data, dict) and isinstance(list(exp_data.values())[0], dict):
+        exp_vals = {}
+
+    n_qubits = get_qubit_number(mapping, ferm_ham.n_spinorbitals)
+
+    # Go over all terms in fermionic Hamiltonian
+    for term in ferm_ham.terms:
+        length = len(term)
+
+        if not term:
+            continue
+
+        # Fermionic term with a prefactor of 1.0.
+        fermionic_term = FermionOperator(term, 1.0)
+
+        qubit_term = fermion_to_qubit_mapping(fermion_operator=fermionic_term,
+                                            n_spinorbitals=ferm_ham.n_spinorbitals,
+                                            n_electrons=ferm_ham.n_electrons,
+                                            mapping=mapping,
+                                            up_then_down=up_then_down)
+        qubit_term.compress()
+
+        # Loop to go through all qubit terms.
+        eigenvalue = 0.
+
+        if isinstance(exp_data, RandomizedClassicalShadow):
+            for qterm, coeff in qubit_term.terms.items():
+                if coeff.real != 0:
+                    # Change depending on if it is randomized or not.
+                    eigenvalue += exp_data.get_term_observable(qterm, coeff)
+
+        if isinstance(exp_data, dict):
+            for qterm, coeff in qubit_term.terms.items():
+                if coeff.real != 0:
+
+                    try:
+                        exp_val = exp_vals[qterm] if qterm else 1.
+                    except KeyError:
+                        ps = pauli_of_to_string(qterm, n_qubits)
+                        hist = aggregate_histograms(*[Histogram(exp_data[basis]) for basis in filter_bases(ps, exp_data.keys())])
+                        exp_val = hist.get_expectation_value(qterm, 1.)
+                        exp_vals[qterm] = exp_val
+
+                    eigenvalue += coeff * exp_val
+
+        # Put the values in np arrays (differentiate 1- and 2-RDM)
+        if length == 2:
+            iele, jele = (int(ele[0]) for ele in tuple(term[0:2]))
+            onerdm[iele, jele] += eigenvalue
+        elif length == 4:
+            iele, jele, kele, lele = (int(ele[0]) for ele in tuple(term[0:4]))
+            twordm[iele, lele, jele, kele] += eigenvalue
+
+    # Construct spin-summed 1-RDM.
+    for i, j in it.product(range(onerdm.shape[0]), repeat=2):
+        onerdm_spinsum[i//2, j//2] += onerdm[i, j]
+
+    # Construct spin-summed 2-RDM.
+    for i, j, k, l in it.product(range(twordm.shape[0]), repeat=4):
+        twordm_spinsum[i//2, j//2, k//2, l//2] += twordm[i, j, k, l]
+
+    return onerdm, twordm, onerdm_spinsum, twordm_spinsum
```

Comment:
Should we put flags to return spinsum RDMs or not?

Reply:
Done